File: libgomp/testsuite/libgomp.fortran/target-map-iterators-2.f90
! { dg-do run }

! Test transfer of dynamically-allocated arrays from target using map
! iterators.

program test
  implicit none

  integer, parameter :: DIM1 = 8
  integer, parameter :: DIM2 = 15

  type :: array_ptr
    integer, pointer :: arr(:)
  end type

  type (array_ptr) :: x(DIM1)
  integer :: expected, sum, i, j

  call mkarray

  !$omp target map(iterator(i=1:DIM1), from: x(i)%arr(:)) map(from: expected)
    expected = 0
    do i = 1, DIM1
      do j = 1, DIM2
	x(i)%arr(j) = (i+1) * (j+1)
	expected = expected + x(i)%arr(j)
      end do
    end do
  !$omp end target

  sum = 0
  do i = 1, DIM1
    do j = 1, DIM2
      sum = sum + x(i)%arr(j)
    end do
  end do

  if (sum .ne. expected) stop 1
contains
  subroutine mkarray
    do i = 1, DIM1
      allocate (x(i)%arr(DIM2))
    end do
  end subroutine
end program

File: libgomp/testsuite/libgomp.c-c++-common/target-map-iterators-2.c
/* { dg-do run } */
/* { dg-require-effective-target offload_device_nonshared_as } */

/* Test transfer of dynamically-allocated arrays from target using map
   iterators.  */

#include <stdlib.h>

#define DIM1 8
#define DIM2 15

void mkarray (int *x[])
{
  for (int i = 0; i < DIM1; i++)
    x[i] = (int *) malloc (DIM2 * sizeof (int));
}

int main (void)
{
  int *x[DIM1];
  int y, expected;

  mkarray (x);

  #pragma omp target enter data map(alloc: x)
  #pragma omp target map(iterator(i=0:DIM1), from: x[i][:DIM2]) \
		     map(from: expected)
    {
      expected = 0;
      for (int i = 0; i < DIM1; i++)
	for (int j = 0; j < DIM2; j++)
	  {
	    x[i][j] = (i+1) * (j+1);
	    expected += x[i][j];
	  }
    }

  y = 0;
  for (int i = 0; i < DIM1; i++)
    for (int j = 0; j < DIM2; j++)
      y += x[i][j];

  return y - expected;
}
